mail_tracking/models/mail_tracking_email.py

```diff
@@ -0,0 +1,497 @@
+                score += weights.get(state, 0.0) * states[state]
+        else:
+            for tracking in self:
+                score += weights.get(tracking.state, 0.0)
+        if score > 100.0:
+            score = 100.0
+        elif score < 0.0:
+            score = 0.0
+        return score
+
+    @api.depends("recipient")
+    def _compute_recipient_address(self):
+        for email in self:
+            email.recipient_address = False
+            recipient_email = email_split(email.recipient)
+            if recipient_email:
+                email.recipient_address = recipient_email[0].lower()
+
+    @api.depends("name", "recipient")
+    def _compute_tracking_display_name(self):
+        for email in self:
+            parts = [email.name or ""]
+            if email.recipient:
+                parts.append(email.recipient)
+            email.display_name = " - ".join(parts)
+
+    @api.depends("time")
+    def _compute_date(self):
+        for email in self:
+            email.date = fields.Date.to_string(fields.Date.from_string(email.time))
+
+    def _get_mail_tracking_img(self):
+        m_config = self.env["ir.config_parameter"]
+        base_url = m_config.get_param("mail_tracking.base.url") or m_config.get_param(
+            "web.base.url"
+        )
+        if self.token:
+            path_url = (
+                f"mail/tracking/open/{self.env.cr.dbname}/{self.id}/{self.token}/"
+                f"blank.gif"
+            )
+        else:
+            # This is here for compatibility with older records
+            path_url = f"mail/tracking/open/{self.env.cr.dbname}/{self.id}/blank.gif"
+        track_url = urllib.parse.urljoin(base_url, path_url)
+        _logger.debug(f"Sending email will tracking url: {track_url}")
+        return f'<img src="{track_url}" alt="" data-odoo-tracking-email="{self.id}"/>'
+
+    def _partners_email_bounced_set(self, reason, event=None):
+        recipients = []
+        if event and event.recipient_address:
+            recipients.append(event.recipient_address)
+        else:
+            recipients = [x for x in self.mapped("recipient_address") if x]
+        for recipient in recipients:
+            self.env["res.partner"].search(
+                [("email", "=ilike", recipient)]
+            ).email_bounced_set(self, reason, event=event)
+
+    def smtp_error(self, mail_server, smtp_server, exception):
+        values = {"state": "error"}
+        IrMailServer = self.env["ir.mail_server"]
+        if (
+            str(exception) == IrMailServer.NO_VALID_RECIPIENT
+            and not self.recipient_address
+        ):
+            values.update(
+                {
+                    "error_type": "no_recipient",
+                    "error_description": "The partner doesn't have a defined email",
+                }
+            )
+        else:
+            values.update(
+                {
+                    "error_smtp_server": tools.exception_to_unicode(smtp_server),
```

Comment:
is smtp_server an exception?

Reply:
yeah u'r right, how silly I am, it's just a name

wonder why the author uses it :sweat_smile: